### PR TITLE
refactor: shard lock module

### PR DESCRIPTION
**Pull request by ShiKaiWi**
## Which issue does this PR close?

Closes #

## Rationale for this change
 There are some issues remaining in the #706.

## What changes are included in this PR?
- Address the comment reviews left in #706.

## Are there any user-facing changes?
None.

## How does this change test
Existing tests.

### CODE REVIEW

File: cluster/src/shard_lock_manager.rs [hunk 1]
```diff
@@ -85,12 +106,59 @@ pub struct ShardLockManager {
...
+    }
+
+    /// Get the duration until the lease is expired.
+    ///
+    /// The actual duration will be still returned even if the lease is not
+    /// alive. And None will be returned if the lease is expired.
+    fn duration_until_expired(&self) -> Option<Duration> {
+        let expired_at = self.expired_at;
+        let now = Instant::now();
+        if expired_at > now {
```
Comment (jiacai2050) [inline cluster/src/shard_lock_manager.rs:136]:
```
expired_at.checked_duration_since(now)
```

File: cluster/src/shard_lock_manager.rs [hunk 2]
```diff
@@ -344,50 +491,104 @@ impl ShardLock {
...
+                            "The lease is about to expire, and trigger the on_lock_expire callback, shard_id:{shard_id}, \
+                             lease_id:{lease_id}"
+                        );
+
+                        on_lock_expired(shard_id).await;
+                        return;
+                    }
+                }
+
+                let (lease_expire_notifier, mut lease_expire_receiver) = oneshot::channel();
```
Comment (jiacai2050) [inline cluster/src/shard_lock_manager.rs:536]: It seems you have used both receiver and rx, better to use only one style here.

File: cluster/src/shard_lock_manager.rs [hunk 3]
```diff
@@ -344,50 +491,104 @@ impl ShardLock {
...
-                    _ = timer => {
-                        if lease.is_expired() {
-                            warn!("The lease of the shard lock is expired, shard_id:{shard_id}");
-                            on_lock_expired(shard_id);
-                            return
-                        }
+                if let Some(dur_until_expired)  = lease_for_bg.duration_until_expired() {
+                    // Assume the keepalive cost is 1.5 rpc_time to ensure the lease is not expired when waiting for next keepalive.
+                    let keepalive_cost = rpc_timeout + rpc_timeout / 2;
+                    if dur_until_expired > keepalive_cost {
```
Comment (jiacai2050) [inline cluster/src/shard_lock_manager.rs:516]:
```
if let Some(wait) = dur_until_expired.checked_sub(keepalive_cost)
```

File: cluster/src/shard_lock_manager.rs [hunk 4]
```diff
@@ -344,50 +491,104 @@ impl ShardLock {
...
+                        let wait_dur = dur_until_expired - keepalive_cost;
+                        info!(
+                            "Next keepalive will be schedule after {wait_dur:?}, shard_id:{shard_id}, lease_id:{lease_id}, \
+                            keepalive_cost:{keepalive_cost:?}, duration_until_expired:{dur_until_expired:?}"
+                        );
+                        tokio::time::sleep(wait_dur).await;
+
+                        info!("Start to keepalive, shard_id:{shard_id}, lease_id:{lease_id}");
+                    } else {
+                        warn!(
```
Comment (jiacai2050) [inline cluster/src/shard_lock_manager.rs:526]: Why do we need to expired in this case?

File: cluster/src/shard_lock_manager.rs [hunk 5]
```diff
@@ -344,50 +491,104 @@ impl ShardLock {
...
+        // procedure.
+        let initial_state = LeaseState::new(expired_at);
+        let lease = Arc::new(Lease::new(
+            lease_id,
+            Duration::from_secs(self.ttl_sec),
+            initial_state,
+        ));
+
         let shard_id = self.shard_id;
+        let rpc_timeout = self.rpc_timeout;
```
Comment (Rachelint) [inline cluster/src/shard_lock_manager.rs:504]: The network jitter tolerating startegy here seems complex, maybe add a detail comment to describe the general procedure?

File: cluster/src/shard_lock_manager.rs (hunk 4 above)
Comment (ShiKaiWi) [reply cluster/src/shard_lock_manager.rs:557]: There is no enough time to do a next keep alive call in the case where the timeout happens.

**Review by jiacai2050 — APPROVED** [review]
LGTM

**Pull request merged**